Support Unicode 16.0

khwilliamson opened the pull request:
This branch adds support for Unicode 15.1, then 16.0

The process for upgrading to new Unicode releases is mostly automatic, and non-problematic., with the exception of the break position algorithms, used in regular expression patterns `\X \b{gcb} \b{wb} \b{lb} \b{sb}`.  `sb` hasn't been changed in a long time, and this PR doesn't affect it.  The other ones are.

The new releases are described in https://www.unicode.org/versions/Unicode15.1.0/
and https://www.unicode.org/versions/Unicode16.0.0/

I assert that no correctly functioning program should expect that an unassigned code point that is eligible for assignment will stay that way  And that is what these releases are mostly about, creating characters where before there were none.

The best case for saying that this pull request changes the behavior of correctly functioning programs is U+5146, which is a Han script character meaning trillion in Taiwan and Japan, and million in mainland China.  Prior to 15.1 the trillion meaning was what Unicode took; starting with 15.1, it takes on the mainland meaning of million.

Other than this, the changes (as opposed to additions) involve the `\X (\b{gcb}`) and especially the `\b{lb}` constructs.  The latter has significant improvements for Indic languages, which Unicode says brings its Standard into much better compliance with what native speakers expect.  There are also changes to the line breaking algorithm for text enclosed in `«»` quotation marks used In France.

This series of commits is lengthy, but all but a few are exclusively about the tables used by `regexec.c` to determine if a break of the given type is permissible at this position in the parse string.  The first many are to improve our infrastructure to make it easier to update these properties in the future.  As a result, the code actually doing the version updates is minimal.  The tables affected by the improvement commits is under source control, and can be seen to be unaffected by them, hence with no behavior change at all, or explained in the corresponding commit message as mostly being bug fixes.

The crux of the first set of commits is to make the code in `mk_invlists' closely resemble the text in the Unicode documents that spell out the rules.  In essence, those documents define pseudo-code.  This makes it far easier to compare what we have with what they say to do.  I have WIP to actually parse the text of the documents, but that is for later.  The documents are  https://www.unicode.org/reports/tr29 and https://www.unicode.org/reports/tr14/.  The culmination of that is to change this file to have the rules here ordered the same way as the documents do.  There are certain advantages to doing things in reverse order, so that the lowest priority rule is applied first, and then overwritten by higher priority rules as needed.  But it made it hard to compare what we have with what Unicode has.  So these commits use the Unicode ordering and do the extra bookkeeping that was avoided by the old ordering.

Most of the rules for deciding if a break is possible don't require context.  You don't break a word or a line between alphabetic characters (in Western languages anyway).  You don't need to look further than the two characters on either side of the current position.  Other pairs require more context to decide.  In "123." do the digits form a word? The answer is maybe.  If the character after the dot is a space, yes they are considered to form a word and the dot is probably indicating the end of a sentence.  If the character after the dot is 4, the dot is part of the word '123.4"  For rules that require context, small DFAs are needed, and are placed in `regexec.c`  Unicode writes its rules assuming that the DFAs are stackable, that at any given position, you can look behind, and if you find a quote you do one thing; if instead you find a parenthesis, you do another things, etc.  Perl got away without having to implement this full generality until 15.1 really made it necessary.  The crux of this branch is changing to use stacked DFAs.  This removed the need for the existing workarounds to compensate for the lack of that generality, nd makes adding new DFAs in the same spot much easier.





* This set of changes requires a perldelta entry, to be furnished

khwilliamson committed:
mk_invlists: Clarify output table headings

Changes the wording for some table headings in the generated file to
indicate where to find what the abbreviations mean

khwilliamson committed:
mk_invlists: White-space comments

This includes outdenting and indenting where future commits will add or
remove blocks

khwilliamson committed:
mk_invlists: Fix rule LB13

This rule was written here to not include the actions when the character
before the candidate break position is a number.  This is just plain
wrong.  The Unicode rules have never said this.

khwilliamson committed:
mk_invlists: Add extensive comments

khwilliamson committed:
mk_invlists.pl: Use feature signatures

khwilliamson committed:
mk_invlists: Move some lines earlier in their functions

where the next commits will want them

khwilliamson committed:
mk_invlists: Set and get break table values with functions

Previously, we would just set an individual element directly.  This
changes most of those to use function calls instead.  This has two main
benefits.  The function can change what's being done without having to
change many lines; and these sets had a lot of visual noise with sigils
and hash references.  The result is a lot easier to read.

The next few commits will continue this process.

Note that the generated tables are unchanged by this commit.  It has no
effect on runtime processing.  That will be true of the next commits as
well.

It became obvious in doing this that the rule for Perl_Tailored_HSpace
does not belong in the 3's, but comes immediately before that.
Arbitrarily use '2z'

khwilliamson committed:
mk_invlists: Reorder two statements

This now matches the order that Unicode gives; for easier checking that
our code matches their demands.

khwilliamson committed:
mk_invlists: Allow arbitrary list of cells

This follows up on the previous commit which allowed simply specifying
an entire row or column.  This adds the ability to specify a list.

khwilliamson committed:
mk_invlists: Add ability to specify an entire row simply

Instead of having to loop through all the cells of a row or column, this
commit uses '*' to represent the whole thing.  This is more in keeping
with the text of the Unicode rules which just leaves thing blank if it
means everything;

khwilliamson committed:
mk_invlists: Add ability to specify a complement of list

And use it in one instance.

Previous commits have added the ability to pass multiple items simply to
the functions that work on rows and columns.  This now gives the ability
to complement the set of the multiple items passed.

khwilliamson committed:
mk_invlists: move decls comments around

khwilliamson committed:
mk_invlists: Handle Combining Mark: changes CMxZWJ

This is separated out from the previous commit because it is tricky XXX

khwilliamson committed:
mk_invlists: Allow abbreviations for break classes

And use them in a couple of places.  This will allow the rules to more
closely align with the Unicode text, which uses abbreviations just
sometimes.

khwilliamson committed:
mk_invlists: Change a word to be more accurate

Everything is an action.  Some are accomplished via DFAs.  This commit
uses the latter word in places where it is a DFA.  It actually uses this
new term where it doesn't apply.  Future commits will remove those
inaccuracies.

khwilliamson committed:
mk_invlists: Add effectively macro expansions

Unicode's Word Break rules have shortcut names that really mean multiple
ones.  For example, AHLetter means either ALetter or Hebrew_Letter.

This commit allows "macros" to be defined like this so that the
statements in this file more closely resemble those of the Unicode text.

More importantly, Unicode's rules in recent times need subdivided
equivalence classes, such as Alphabetics that are also East Asian.  What
has been done so far is when that happened, extra rules were added that
were all possible combinations of these subdivisions.  It is easy to
miss a combination; and it turns out there are bugs.  This new
capability allows us to say that an Alphabetic (ALetter) is a
combination of plain ALetters plus East Asian letters, and the code
generates all the combinations automatically.  This makes the text
cleaner and safer.

khwilliamson committed:
mk_invlists: Use abbreviations for Line Break

Unicode UAX #14 gives rules for the Line Break property using the short
names for them.  Prior to this commit, we mostly used the full names for
the classes in this property.  This commit changes to use the short
names.  This makes it easier to compare the code here with the UAX text.
The abbreviations aren't always straight forward, so it was easy to go
astray.

khwilliamson committed:
mk_invlists: Improve DFA names

This commit now imposes more structure on the names.

The names are sort of pseudo code that lays out what the DFA is to do.
The most significant change is to standardize what has been done in
recent commits with newly added DFAs.  And that is to use the string
'_v_' in the name where the tip of the 'v' points to where position in
the input string being processed where this rule applies to.

khwilliamson committed:
mk_invlists: Fix rule LB11

This rule is not affected by spaces, yet the code was saying it should
be.

khwilliamson committed:
mk_invlists: Add fields to data structure

This converts each DFA form just a number into a separate hash in a
bigger hash with more information besides that number.

This extra information will be needed in a future commit.

khwilliamson committed:
mk_invlists: Change two formal parameter names

khwilliamson committed:
mk_invlists/regexec.c: Generate and use macros

With this commit, mk_invlists.pl now generates #define macros isFOO that
regexec.c now uses to determine if a character is in a particular line
breaking class.  Previously, x == foo was used.  This change insulates
the code from having to worry about when classes get changed to be
combinations.

khwilliamson committed:
mk_invlists: Use new split capability with AHLetter

The description in UAX #29 of Unicode's Word Break property uses two
convenience macros to simplify some of their rules.

The split capability introduced several commits ago, allows this program
to follow along, making the rules here more closely aligned to the text
in UAX 29, hence simpler.

This commit creates one macro, AHLetter; the next commit does the other
macro.

The name of the DFAs involving this name are changed to correspond.

khwilliamson committed:
mk_invlists: Use new split capability with MidNumLetQ

This follows on the previous commit, with the other Word Break property
name that Unicode macroizes

khwilliamson committed:
mk_invlists: Fix rule LB12

This rule is not affected by spaces, yet the code was saying it should
be.

khwilliamson committed:
mk_invlists: Add rule numbers to break table output

This is very helpful in debugging, and correlating the tables with the
Unicode UAX documents from which they are derived.

khwilliamson committed:
mk_invlists: Narrow some output tables

Future Unicode releases will greatly explode the size of certain tables.
Prior to this commit, the minimum column size was two, but some table
columns fit in a single window column.  This commit changes to use the
minimum required.

khwilliamson committed:
mk_invlists: Center row labels in output tables

This improves readability

khwilliamson committed:
mk_invlists: Improve output table column headings

This uses a more complex algorithm to generate short labels to demarcate
rows and columns in some output tables.

This doesn't affect the current tables for Unicode 15.0, but will in
future Unicode releases.

khwilliamson committed:
mktables: Consolidate code into a single function

Some properties in Unicode essentially form equivalence classes for all
possible code points.

For example, Unicode publishes the Line Break (LB) property, where each
possible code point is given a type, like Alphabetic, or Opening
Parenthesis.  All code points that act as alphabetics have the AL
equivalence class.  All that act like Opening Parentheses have the OP
class.

Unicode also publishes rules as to if it is permissible to break between
code point of any types.  For the Line Break property, you wouldn't
break a line between two alphabetics or between an opening parenthesis
and an alphabetic, but you could between a Space and almost any other
type or between a closing parenthesis and many types.

Perl uses these properties to implement the \b{lb} etc regular
expression constructs.  It uses a two-dimensional array where the value
in the cell [x,y] tells whether a break is permissible between
characters of type x and characters of type y.  (Some cases can't be
done with this simple lookup, but knowing the surrounding context is
necessary to make a decision.  Those are implemented as DFAs in
regexec.c.)

Unicode used to publish such an array for the Line Break property, and
still publishes some non-normative .html files that contain similar
information.  But to really know what to do, one has to read documents
UAX#14 and UAX#29 that contain textual descriptions of the rules.  These
change each new release, and are the major pain in upgrading to a new
release.

In recent releases, Unicode has mostly stopped creating new equivalence
classes as it has refined the rules for the boundary conditions  For
example, the line boundary conditions are very different for East Asian
(EA) characters than the Western scripts.  Effectively there are thus
two sets of rules.  But instead of creating new equivalence classes that
reflect this reality, Unicode has chosen to just document it in those
two UAX documents.  I don't know the motivation for this.

But perl wants that table to divvy up all the possible boundary
conditions, so it can continue to use the array to make most of the
decisions, so mktables splits the equivalence classes that Unicode
provides into new ones that reflect what the UAXes say.  At first, I
thought this was a one-off matter, so wrote a few lines to handle a
special case; then when the next release came out, added a few more for
another one, etc.  But Unicode 15.1 and 16.0 continue the trend, so it's
become an effort.

This commit consolidates the previous one-off code snippets into one
generalized function.  It should be able to handle future instances
without having to craft something new each time.

It also creates a new data structure that mk_invlists.pl can look at so
that it doesn't have to repeat the logic found here, as it currently
does.

khwilliamson committed:
mk_invlists: Add no_nobreak_override()

This new function allows removing loops from the main code

khwilliamson committed:
mk_invlists: Add ability to get set subtraction

This capability will be used in future commits, so that the
implementation can more closely follow Unicode's text

khwilliamson committed:
mk_invlists: Use new set subtraction ability

This allows the removal of some combinatorial complexity, thus showing
a bug in which the combination of PO to EOP had not been added when it
should have been.

Currently, mktables splits the Line Break OP and CP classes into East
Asian ones, and the remainders.  The extra combinations occurred because
the code here needed to take every existing OP and add an East_Asian
(EA_OP) equivalent; same with CP.  It's easy to miss one, and I did.

This commit allows this split to be hidden from most places in
mk_invlists.

khwilliamson committed:
mk_invlists: Hoist calculation to sub callers

And pass the result to the subroutine.

This is in preparation for this value to be needed in additional places.

khwilliamson committed:
mk_invlists: Set values in unused table cells to 0

These cells exist so that code is less likely to need to be changed when
a new Unicode release comes along.  Currently it doesn't matter at all
what is in those cells, because they are never read.  But future commits
will want to make sure they don't refer to dfas that are obsolete and
whose references to could be undefined symbols that would abort the
compilation.

The choice of 0 or 1 to put in the cells was arbitrary; I know of no
reason to prefer one or the other

khwilliamson committed:
mk_invlists: Move break tables edge cells

These tables are placed in charclass_invlists.h.  They have a row and
column for what happens when the position being checked for is at the
start or end of the text.  This commit reorders the tables so that the
edge row and column are, well, at the edges.  And it relabels the labels
to be '^' and '$' respectively.

khwilliamson committed:
mk_invlists: Use new split capability with ExtPict

\p{Extended_Pictographic} is not fully implemented yet because unlike
other properties, it can match a string instead of a single character.

And it is kind of a kludge here  The 14.0 release was analyzed by me and
the rules here were customized based on that analysis.  For example, in
the Line Break property, a clause was added by Unicode to Rule LB30b
that required taking the intersection of this property and all the
Unassigned code points.  It turns out that everything in that
intersection had the Line Break class of Ideographic, so I modified
mktables to split the Ideographic class into two components, the
elements of the intersection went into the long-named
"Unassigned_Extended_Pictographic_Ideographic" and plain Ideographic was
left with the remainder.  To match all of Ideographic you have to
specify both classes.  By using the new split capability, this can be
done effectively as a macro expansion, and the special cases can be
removed from the code.  This commit does this.

Similarly, both the Word Break and Grapheme Cluster Break properties
have somewhat different interactions with Extended_Pictographic that
this commit smooths over.

This situation is brittle.  A new release of Unicode might change things
so that Ideographic isn't the only LB class in the intersection
mentioned above, so the customization has to be checked in every
release.  A few commits later in this branch, this will be automated,
and no longer a concern.

khwilliamson committed:
mk_invlists: Remove obsolete function

This function was used when the previous scheme of applying the rules in
reverse order needed to be overridden in a few cases by prohibiting
changes to existing seemingly lower priority values.  Now there's
no lower priority value in the cell that we would need to preserve.

khwilliamson committed:
mk_invlists: Use 'for' statement modifier

This significantly cuts down on the verbiage, and makes the rules in
this file more closely match the text from which they are derived in UAX
14 and UAX 29

khwilliamson committed:
mk_invlists: Reverse order of break property rules

Before this commit, the rules for populating the tables for break
properties were laid out in reverse order, so that the lowest priority
rule was executed first.  It filled a cell, which then would be
overwritten by any higher priority rule that applied to it.

This reverse order made it harder to compare the rules with the text of
the Unicode rules these are trying to implement.

This commit changes things to have the rules in the same order as
Unicode lists them.

The previous scheme had certain advantages that this has to make up for
by using temporary code to override what would otherwise have gone into
the cells.  This code will no longer be needed in a few commits when a
general purpose stacking DFA scheme is implemented.

As a result, of this temporary code, only two cells in one property
change as a result of this complete reversal.  They change to using a
DFA which ends up returning the same results as the original
unconditional value.

khwilliamson committed:
mk_invlists: Add ability to tie table cell changes together

Some Unicode rules say that some cells are supposed to be changed at the
same time their master cells are.  This commit forms that linkage.

khwilliamson committed:
mk_invlists: rm no longer used sub

The previous commit took away the need for this.

khwilliamson committed:
mk_invlists: Remove some special cases

These were added to compensate for reversing the order of handling the
break property rules.  This commit hides the need for that in one place
per table, except for a second place for Line Break.

The only changes to the tables occur in the garbage row and column which
aren't actually accessed, so those changes are harmless.

It is a temporary commit.  A few commits from now, this will be removed.

khwilliamson committed:
mk_invlists: Use new mktables enhancements

Now mk_invlists no longer has to know what the details are of properties
that have been split into more, smaller equivalence classes.  mktables
handles that and provides the information in new hashes.

khwilliamson committed:
regexec.c: Change static function API

This makes it clearer to use.  Instead of having a boolean flag to
change the behavior, there are now two macros that call the underlying
function, and their names reflect the desired behavior

khwilliamson committed:
regexec.c: Change function name

The new name is longer, but it makes clear that it does something that
the reader of this code might find unexpected.

khwilliamson committed:
mk_invlists: Look for a DFA optimization possibility

If both branches of an else lead to the same result, skip the else and
set the result unconditionally.  That's what this commit does for DFAs
that get the same value if they succeed as when they don't.

There is one current case where the DFA can return an anomalous result,
so it can't be optimized out.  Add a field to the hash entry defining
that entry, so it doesn't get optimized.

khwilliamson committed:
mk_invlists: Reorder some hash entries

The new order is based on the order of their respective rules; the next
commit expands these, and it makes it easier for a human to look up.

khwilliamson committed:
mk_invlists: Change doubled semicolon to single

khwilliamson committed:
mk_invlists: Generalize to stack DFAs for break properties

The Unicode breaking algorithms are supposed to be implemented by
executing DFAs in priority order, stopping at the first one that
succeeds.  (In many cases a DFA isn't needed, and we can unconditionally
say that there is or isn't a break at a given position simply by looking
at the characters on either side of it.)

But it was a significant amount of work to get from where perl started
to be able to do that.  And it hasn't been necessary until now.  In most
cases, a single DFA suffices, and where not, a more complicated single
DFA took care of the stacking.

But this has become untenable in Unicode 15.1, so I ended up doing the
work to implement their algorithm.  The result is more, but simpler
DFAs, and it becomes easier to add new ones, as they don't have to
interact with other ones.  The stacking does that for them.

This commit implements a separate DFA table beyond the x,y lookup table.
If the decision that this is a breakable position requires a DFA, the
x,y contents are an index into this separate table, which contains the
DFA to follow.  The first element gives the case statement number to use
to execute the DFA.  The second element gives the value to return if the
DFA succeeds.  If it fails, the code add +2 to get the next thing to
try.

khwilliamson committed:
mk_invlists: Remove a temporary work-around

This code was due to a few commits ago having reversed the ordering the
Unicode rules are applied in.  After updating to use a generalized DFA
scheme, it is no longer needed

khwilliamson committed:
mk_invlists: Remove hard-coded numbers

A couple of commits ago, the last necessarily-hard-coded DFA enum
besides 0 and 1  was removed.  This allows for all the rest to be
assigned by using the value of an incrementing variable.

This makes it easy to add DFAs in the middle of existing ones, as will
happen as future Unicode releases come our way.

khwilliamson committed:
mk_invlists/regexec.c: Prepare for Unicode 16.0

khwilliamson committed:
mk_invlists: Include cells in calculating column widths

This program generates tables for the Break properties that are somewhat
human readable.  Before this commit, just the heading line for a column
determined its width.  This commit factors in the maximum width of any
cell in the column as well.  It used to be that this required a separate
pass, and so wasn't done.  But now that separate pass is required anyway
for other reasons, and it is simple to add to it this check.

khwilliamson committed:
mktables: Prepare for Unicode 16.0

Commit:
Add Unicode 16.0

This is includes updates to a few perl files that need to know the
current Unicode version, and regenerating perl files that depend on the
Unicode data

khwilliamson committed:
mktables: Note break table code for Unicode 16.0 is updated

khwilliamson committed:
mk_invlists: Sort some lists

These lists are densely packed.  It is easier to find something if they
are sorted

khwilliamson committed:
mk_invlists: Remove a no longer used enum

The new generalized DFA scheme makes this value moot; it was used to get
around not having such a scheme.

khwilliamson committed:
mk_invlists: Remove no longer used function

khwilliamson committed:
mk_invlists: Restore calculation of new keywords, etc

Now we are ready to use a new Unicode version, we have to regenerate
everything.  This was turned off earlier in this branch temporarily
until now so as to speed up the testing, as it was known these values
wouldn't change until now.

khwilliamson committed:
mk_invlists: Use new split capability with ALetter

ALetter also contains the class ExtPict_LE.  Prior to this commit, there
had to be a rule for each ALetter doing the same thing with ExtPict_LE.
But the new splits capability allows ALetter to expand automatically to
both.

This uncovers a bug.  There should have been a rule
 WB5 ALetter x ExtPict_LE
which was missing.

khwilliamson committed:
regexec.c: Change static function API

Sometimes this functionality is needed to also skip over certain
intervening classes of characters while backing up in the parse string.
This commit creates two macros to call the modified underlying function
with a boolean flag.  This names of the macros make it easy to know
what's happening.

khwilliamson committed:
regexec.c: Skip CM and ZWJ in look behind in LB parsing

The Unicode standard says that these two characters are to be ignored
for the purposes of determining if there is a Line Break just before
certain characters.  That is, you have to back up in the parse string
past all adjacent ones of these, and then examine it.

This applies to any lower priority rule than LB9.  This commit fixes two
cases that didn't do that.

khwilliamson committed:
mk_invlists: Use new DFA scheme for horizontal white space

Perl doesn't follow the Unicode standard with regard to its treatment of
white space, in particular sequences of horizontal white space.  Unicode
allows "tailoring" of its rules for local situations, and Perl
traditionally with \B  has treated all sequences of white space as a
single unit.  Unicode originally considered each space in a sequence of
them as a separate unit.  A perl program would want them all a single
unit.  Unicode eventually came round to our way of thinking, but not
entirely, as comments unaffected by this commit indicate.

The DFA for this situation does not fit in with the new stackable DFA
scheme, and woul start failing tests a few commits later as the shim
code is removed.  Convert to the new scheme, which allows us to call the
functions that affect a single cell twice with effect.  The order is
immaterial, but one call installs a default behavior, and the other a DFA
that ends up being executed first to override that behavior in certain
(rare) situations.

khwilliamson committed:
mk_invlists: Remove accesses of unused parameter

set_cells() no longer reads this parameter; no need to pass it nor set it
up.

khwilliamson committed:
mk_invlists: Add a shorter form DFA

This is just for legibility of reading the rules

khwilliamson committed:
mktables: Prepare for Unicode 15.1

khwilliamson committed:
Temporarily skip regen porting test in this branch

The digest numbers keep changing in this branch.  Turn this test off
until near its end.

khwilliamson committed:
mk_invlists: Temporarily don't generate some porting info

This series of commits has dozens of commits that would otherwise
require much more work to generate.  This commit temporarily turns off
generating EBCDIC tables, and the tables that only change when a new
Unicode release happens.  Bisecting on an ASCII machine is unaffected

khwilliamson committed:
regexec.c: Rename a couple of variables

The previous names erroneously implied these were associated with the
parameters to these functions; instead rename to indicate they are
associated with some local variables.

khwilliamson committed:
mk_invlists: Add stack trace facility

This is useful in debugging

khwilliamson committed:
lib/Unicode/UCD.t: Prepare for Unicode 15.1

The  numeric value for U+5146 changed in 15.1

khwilliamson committed:
mk_invlists/regexec.c: Prepare for Unicode 15.1

khwilliamson committed:
mk_invlists: Restore generating EBCDIC

This had been turned off in this branch to speed up compilatian, and
hence development.  The code mostly changed in this branch is the same
as in ASCII anyway.  It could have become an issue only if someone tries
to bisect on an EBCDIC machine, which I don't believe has happened, if
ever, in decades.

khwilliamson committed:
Revert "Temporarily skip regen porting test in this branch"

This temporary commit has now served its purpose.

haarg commented:
I think it would be a shame to ship a new perl release that didn't have the latest Unicode standard. Something breaking based on a new Unicode version seem unlikely to be found in things like BBC reports. It seems more likely to only be found in a stable release. And should only impact programs that were already doing something incorrect. So I'm inclined to go ahead with merging this PR before 5.42.

However, this is a rather large change to `regexec.c` which I'm not sure I'm the best person to evaluate. And this PR has a failing test.

Nit: The "Temporarily skip" commit and its revert should be removed before merging.

Leont commented:
> 
```
Failed test ''10000000000000000' is listed as an alias for prop_value_aliases('-_ nv', '1_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0')'`
```

It is suspicious that a test involving a number that doesn't fit in a 32 bit integer fails on a 32 bit perl.

jkeenan commented:
This p.r. is repeatedly failing one test in the 'linux i386/ubuntu' run on our GH CI.
```
#   Failed test ''10000000000000000' is listed as an alias for prop_value_aliases('-_ nv', '1_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0')'
#   at ../lib/Unicode/UCD.t line 1289.
# 
# Looks like you failed 1 test of 15362.
# 
lib/Unicode/UCD .................................................. FAILED at test 7518
```
I re-started that particular run, but am observing the same error as in the original run.

khwilliamson committed:
UCD.t: Skip test which fails on 32 bit words

In Unicode 15.1, the ideograph U+4EAC now has a numeric value, and that
value is 10 quadrillion (1e+16).  This is the first instance in Unicode
of an integer not fitting in a 32 bit word, as this requires 49 bits.
One of the tests in UCD.t requires round-trip equality in converting
from string to number and back; skip it for this case and any future
similar ones.

I find it interesting that U+4EAC is listed as having the meaning
"capital city".

khwilliamson committed:
mk_invlists: Update comments

khwilliamson committed:
perldelta for Unicode update

khwilliamson commented:
The character '9' has the numeric value 9.  We're used to characters only taking on numeric values of 0-9.  But some scripts have characters that have numeric values be something else, typically when the script doesn't use decimal positional notation. A fairly common case is the script will have a a symbol for 10, one for 20, etc.  Roman numerals are a case which many people in the West know about that have single characters that signify other than 0-9.  (I recently learned that it wasn't until Shakespeare's time that England taught school children anything but Roman numerals.)  Some scripts, typically East Asian, have characters, typically ideographs, that symbolize very large numbers.   Unicode 15.1 introduced for the first time one that has a value that doesn't fit in a 32 bit word.  The build that was failing was a 32 bit one.  And the test that was failing is just one that introduces underscores between digits and verifies that the underscores are ignored, as Unicode requires, in pattern matches.  The test is simplistic, and expects that a round trip of string to number and back yields the same string.  That won't be true for a value that doesn't fit in a word, it is converted to an NV.  So the test isn't valid on such a value, and the fix was to simply skip it.

karenetheridge commented:
Awaiting signoff from a PSC member for this to be in 5.41.11.

khwilliamson commented:
With a bit of private encouragement, I merged this to get it in 5.42.  @karenetheridge  said she would issue a new release should problems with this arise.
